### ISSUE TITLE
Test Stub Does too much

### ISSUE THREAD

**Issue by sol-myob** [monajena27/Toy-Block-Factory#8]
https://github.com/monajena27/Toy-Block-Factory/blob/49a364f7e6039d2b1178996a5e9b2f91c94a6959/ToyBlockFactoryTests/TestPricingCalculator.cs#L9

this thing has way too much logic in it. usually, we try to make the test doubles as dumb as possible. it doesn't matter that it doesn't match the actual pricing calculator. 

for example, GenerateLineItems can just return something like

```foreach block, new lineItem ("Block", 1, 1, 1);```

with no private methods or setup (if you need it to return something specific that can be passed via constructor)

you'll know if your invoice report tests are good if you can make this change fairly easily. The logic that is being tested in the invoice generator shouldn't rely on *what* the prices are, just that they were used,.

I'd also put this in a folder called `testdoubles `or something

**Comment by mona-jena** [comment]
so you just return something for one particular order? How would that test whether the price Calculator works?

**Comment by sol-myob** [comment]
with the price calculator tests! :)

**Comment by mona-jena** [comment]
no I mean if you remove all the logic of how the price calc works from the stub, then how does the test, test that it can calculate if I'm just going to return back what the test wants?

**Comment by sol-myob** [comment]
the idea is to segment and isolate in a way where a failure isn't a house of cards. this is best done along the lines of abstraction where decoupling is high.

We can be sure the price calculator works
We need to be sure that the invoicegenerator works with *any* given implementation of IPricing. therefore we can create the simplest pricing strategy possible to minimize its effect on the invoicecalculator tests (you can think of this in a similar way to, we don't run unit tests with the customer choosing the values).

now with both of those tested. we know we can put them together without concern. to make sure they are put together at all, there is that overall acceptance test which tests your whole factory does what a factory should

**Comment by sol-myob** [comment]
the tests here aren't concerned with testing pricing calculation. they are only concerned with invoice generation. it can be a pitfall testing Auxillary functionality and it is quite the art thinking about what *are* we testing here. what do we want to demonstrate and is this the right place to show that

**Comment by sol-myob** [comment]
wiki does a good job here:
https://en.wikipedia.org/wiki/Test_stub

**Comment by sol-myob** [comment]
https://en.wikipedia.org/wiki/Method_stub

this one is ok too

**Comment by mona-jena** [comment]
But this would only work if you were using it for same order right? I currently have 2 types of order in my InvoiceReportTests, but they would both expect different line items. So how do I give a fixed answer?

**Comment by mona-jena** [comment]
so do we never want to test if priceCalc calculates correctly? Bec if all tests use the stub, then we will never know right if in the actual program which actually calc it, that it works? Ie how do test if the price calc behaves or calculates like we want it to?

**Comment by sol-myob** [comment]
> But this would only work if you were using it for same order right? I currently have 2 types of order in my InvoiceReportTests, but they would both expect different line items. So how do I give a fixed answer?

Good question!

if different arrangements or return values of a dependency are required to exercise a certain capability of the thing we're testing, we can create and use as many different stubs as we need to do that. this can get painful or tedious of course. so usually we craft the stub to either have properties we can change per test or with a constructor that takes the data it needs and then construct the stub for each test with that data.

So using blackjack as an example. we have a random deck that is used for the normal game. for testing purpose we use a deck stub. for each test we `new DeckStub([ace, king, 2])` and then when the program access the `Deal()` method of the stub (via the interface defined method) it will predictably give us the cards we need

> so do we never want to test if priceCalc calculates correctly? Bec if all tests use the stub, then we will never know right if in the actual program which actually calc it, that it works? Ie how do test if the price calc behaves or calculates like we want it to?

we sure do ! but we only need those in the `PricingCalculatorTests.cs` .there should live everything we need to be convinced the pricing calculator does things correctly. when the program runs for real, it is used as the pricing strategy and everything will run swimminly



and if that pricing ever changes. we just need a new version, with new tests just for that version and everything else remains unchanged and tests continue to work as intended

**Comment by mona-jena** [comment]
but PricingCalculatorTests.cs is using the stub rn, should it be using the real one to see if that calculates properly? But we wrote those tests for the stub didn't we?

**Comment by mona-jena** [comment]
this looks so hacky now https://github.com/monajena27/Toy-Block-Factory/blob/fixes/ToyBlockFactoryTests/TestDoubles/TestPricingCalculator.cs

Should I have a different stub file for each order, is that better?

**Comment by sol-myob** [comment]
it sure does. how about
`new PricingCalculatorStub([{"Square",1},{"Circle",2}])` for one test
`new PricingCalculatorStub([{"Square",1},{"Circle",2},{"Some random surcharge",1])` for another

you may need to move the creation of the thing under test out of the test constructor, but that is ok. and it makes sense if each test is created different

when taking in the arguments to your stub remember all you really care about is the description and the qty probably.  you can just hard code the values for the others


so {"Square",3} becomes a square qty 1, price 3, total 3

(my brackets are just matching, not proper c#)

**Comment by mona-jena** [comment]
Don't you want to try and keep the implementation and calling similar to the actual? 

Also in the tests, it gets the order and passes the Blocklist. So doing it this way will mean I will need to change my tests to pass each index of blockList. Basically, break TDD to make it work this way.

**Comment by sol-myob** [comment]
This one is up to you. You can leave it as it is just fine. the constructor of the TestPricing is just a part of the test setup. and then when the pricing methods are called in the implementation you can get those predicatable results.

it's a way to lcoalise the data to each test. atm what you have has all of it together and it is hard to reason about what that test stub looks like it does inside


I am not sure what you mean by breaking TDD. tdd = test driven development

do you mean the red green refactor cycle is broken ?

**Comment by mona-jena** [comment]
cuz i will have to change my current tests to pass in each block instead of blockList which is changing the test cuz of implementation right, though here it is for making the stub work this way

**Comment by sol-myob** [comment]
I understnad, just giving you options

**Comment by mona-jena** [comment]
so this way is ok to do, even though that ^ happens?

**Comment by sol-myob** [comment]
yes, its just odd is all

**Comment by mona-jena** [comment]
I don't like this way either, but don't want to change tests to make it work either, that seems wrong. I could just create multiple stubs for each test if we want to know what the test does and get rid of the case statements? It'll just be a class not doing much that's all.

**Comment by sol-myob** [comment]
yep I agree, the bang for buck on making the change isn's close to being there
